### PR TITLE
Add QSD support

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:2116691
Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
Passed test with test code in https://docs.google.com/document/d/1ucf2A0jbc56EoQuQg8fJyzI-pWVB6gFSCYm3hBmvSFo/edit#
It also depends on https://github.com/avocado-framework/avocado-vt/pull/3468

**Comment by qingwangrh** [comment]
@YongxueHong @zhencliu @hellohellenmao @aliang123 Please help to review.thanks

File: provider/qsd.py [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+    vubp.set_aid(dev_params["id"])
+    devices.append(vubp)
+    LOG_JOB.info("create_vubp_devices %s %s", img_name, devices)
+    return devices
+
+
+def insert_vubp_devices(vm, img_name, params, bus=None):
+    """Insert vhost-user-blk-pci into VM internal devices"""
+    devs = create_vubp_devices(vm, img_name, params, bus)
+    vm.devices.insert(devs)
```
Comment (aliang123) [inline provider/qsd.py:113]: Not sure, why we need this function? When and how can we call it?

We've already have vm.devices.insert(qsd) in create_vubp_devices.

Comment (qingwangrh) [reply provider/qsd.py:113]: you are right,will be removed

File: provider/qsd.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        if self.has_option('--daemonize'):
+            LOG_JOB.info("--daemonize")
+            self.caps.set_flag(Flags.DAEMONIZE)
+        if self.has_option('--pidfile'):
+            LOG_JOB.info("--pidfile")
+            self.caps.set_flag(Flags.PIDFILE)
+
+    def get_pid(self):
+        """ Get QSD pid"""
+        if self.daemon_process:
```
Comment (aliang123) [inline provider/qsd.py:287]: self.daemon_process or self.daemon_process() is right?

Comment (qingwangrh) [reply provider/qsd.py:281]: In my understanding ,self.daemon_process it get address of function, self.daemon_process() it get return of function,
self.daemon_process().get_pid()  TypeError: 'Expect' object is not callable

@YongxueHong ,could you please explain more ?

Comment (YongxueHong) [reply provider/qsd.py:282]: > In my understanding ,self.daemon_process it get address of function, self.daemon_process() it get return of function, self.daemon_process().get_pid() TypeError: 'Expect' object is not callable
> 
> @YongxueHong ,could you please explain more ?

Hi @aliang123 

Here the `daemon_process` is the property of the class, it is not a method that is a callable object. but as a property attribute.
```
    @property
    def daemon_process(self):
        """Return daemon process property."""
        return self._daemon_process
```

 You can consider it likes `foo.a` as the following:
```
class Foo(object):
    self.a = "foo"

foo = Foo()
print(foo.a)
```

File: provider/qsd.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+    def is_daemon_alive(self):
+        if self.daemonize:
+            check_pid_cmd = "ps -q %s" % self.pid
+            if self.pid:
+                return True if process.system(check_pid_cmd, shell=True, ignore_status=True) == 0 else False
+            else:
+                return False
+        else:
+            return super(QsdDaemonDev, self).is_daemon_alive()
```
Comment (YongxueHong) [inline provider/qsd.py:388]:
```suggestion
        return super(QsdDaemonDev, self).is_daemon_alive()
```

File: provider/qsd.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+
+        monitor = qemu_monitor.QMPMonitor(self, self.name, self.qsd_params)
+        monitor.info_block()
+        self.monitor = monitor
+
+    def is_daemon_alive(self):
+        if self.daemonize:
+            check_pid_cmd = "ps -q %s" % self.pid
+            if self.pid:
+                return True if process.system(check_pid_cmd, shell=True, ignore_status=True) == 0 else False
```
Comment (YongxueHong) [inline provider/qsd.py:384]: It looks like too long

File: provider/qsd.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+from virttest.qemu_capabilities import Capabilities
+from virttest.storage import get_image_filename
+
+LOG_JOB = logging.getLogger('avocado.test')
+
+
+class Flags(object):
+    """ Enumerate the flags of QSD capabilities. """
+    DAEMONIZE = 1
+    PIDFILE = 2
```
Comment (YongxueHong) [inline provider/qsd.py:27]:
```suggestion
from enum import Enum, auto

class Flags(Enum):
    """ Enumerate the flags of QSD capabilities. """
    DAEMONIZE = auto()
    PIDFILE = auto()
```

File: provider/qsd.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+            qsd_name = qsd
+            break
+    return qsd_name
+
+
+def create_vubp_devices(vm, img_name, params, bus=None):
+    """Create vhost-user-blk-pci relevant devices"""
+    qsd_name = get_qsd_name_by_image(img_name, params)
+    if not qsd_name:
+        raise ValueError("Can not find QSD")
```
Comment (YongxueHong) [inline provider/qsd.py:65]: We can new an exception for this module:
```
class QsdError(Exception):
    """Generic QSD Error."""
    pass
```

File: provider/qsd.py [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    devs = devs[::-1]
+    for dev in devs:
+        cache_dev = vm.devices.get_by_qid(dev.get_qid())
+        if len(cache_dev):
+            vm.devices.simple_unplug(cache_dev[0], vm.monitor)
+        else:
+            # Force unplug device by raw QMP command
+            if not dev.verify_unplug(None, vm.monitor):
+                out = dev.unplug(vm.monitor)
+                utils_misc.wait_for(lambda: dev.verify_unplug(out, vm.monitor) is True, first=1, step=5, timeout=20)
```
Comment (YongxueHong) [inline provider/qsd.py:133]: It is a little long.

File: provider/qsd.py (hunk 3 above)
Comment (qingwangrh) [reply provider/qsd.py]: updated

File: provider/qsd.py (hunk 4 above)
Comment (qingwangrh) [reply provider/qsd.py]: updated

File: provider/qsd.py (hunk 5 above)
Comment (qingwangrh) [reply provider/qsd.py]: updated

File: provider/qsd.py (hunk 7 above)
Comment (qingwangrh) [reply provider/qsd.py]: updated

File: provider/qsd.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+            raise QsdError("Unknown export type %s " % img["export"]["type"])
+
+        # Prepare image
+        image_created = False
+
+        if name in params.get("images").split():
+            if params.get("force_create_image") == "yes":
+                image_created = True
+            elif params.get(
+                    "create_image") == "yes" and not storage.file_exists(
```
Comment (zhencliu) [inline provider/qsd.py:263]: when QSD inits, the preprocess of VT should have been finished, right?
Then, storage.file_exists should always returns True here

File: provider/qsd.py [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+        self.images = {}
+        self.daemonize = False
+        self.pidfile = None
+        self.pid = None
+
+    def _remove_images(self):
+        for img in self.images.values():
+            if img["image_object"]:
+                LOG_JOB.info("QSD ready to remove image:%s", img["name"])
+                if self.qsd_params.get("remove_image_%s" % img["name"],
```
Comment (zhencliu) [inline provider/qsd.py:210]: 1. if users dont want to remove all images after test, they can set a global var remove_image = no, but here, the image could also be removed even remove_image = no.
2. here we only handle the images created by QSD module, do we need to consider the images created by VT? i.e. consider all images in self.images.

File: provider/qsd.py [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+        if self.has_option('--daemonize'):
+            LOG_JOB.info("--daemonize")
+            self.caps.set_flag(Flags.DAEMONIZE)
+        if self.has_option('--pidfile'):
+            LOG_JOB.info("--pidfile")
+            self.caps.set_flag(Flags.PIDFILE)
+
+    def get_pid(self):
+        """ Get QSD pid"""
+        if self.daemon_process:
```
Comment (zhencliu) [inline provider/qsd.py:300]: refer to line370 line 372, if self.daemonize, self.daemon_process should be None? 
ie. it's possible to set None to self._daemon_process, at that time, use popen to create a daemonized process(line370)

File: provider/qsd.py [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+    def _destroy(self):
+        # Is it already dead?
+        if not self.is_daemon_alive():
+            LOG_JOB.info("dead")
+            return
+
+        if self.daemonize:
+            pid = self.pid
+        else:
+            pid = self.daemon_process.get_pid()
```
Comment (zhencliu) [inline provider/qsd.py:422]: pid = self.get_pid() here?

File: provider/qsd.py (hunk 10 above)
Comment (qingwangrh) [reply provider/qsd.py:300]: updated

File: provider/qsd.py (hunk 9 above)
Comment (qingwangrh) [reply provider/qsd.py:210]: 1. Change to support remove_image = no.
2. To be consistent it is better QSD only handle the images are created by it.

File: provider/qsd.py (hunk 8 above)
Comment (qingwangrh) [reply provider/qsd.py:263]: updated

File: provider/qsd.py (hunk 2 above)
Comment (aliang123) [reply provider/qsd.py:299]: Got it, thanks @YongxueHong

**Comment by aliang123** [comment]
Acked-by: Aihua Liang <aliang@redhat.com>

**Review by hellohellenmao — APPROVED** [review]
LGTM

**Comment by qingwangrh** [comment]
Test example may refer to https://github.com/autotest/tp-qemu/pull/3450

**Comment by zhencliu** [comment]
Discussed with @qingwangrh , the docstrings for image management as well as the qsd cfg samples will be added later,

LGTM

File: provider/qsd.py [hunk 12]
```diff
@@ -0,0 +1,500 @@
...
+        self.monitor = monitor
+
+    def is_daemon_alive(self):
+        if self.daemonize:
+            check_pid_cmd = "ps -q %s" % self.pid
+            if self.pid:
+                return process.system(check_pid_cmd, shell=True,
+                                      ignore_status=True) == 0
+            else:
+                return False
```
Comment (YongxueHong) [inline provider/qsd.py:440]:
```suggestion
            return False
```

File: provider/qsd.py [hunk 13]
```diff
@@ -0,0 +1,500 @@
...
+        "nbd-server": {"addr": {}},
+        "export": {"type": "", "id": "", "node-name": "", "writable": True},
+        "image_object": None
+    }
+
+    def __init__(self, name, params):
+        qsd_params = params.object_params(name)
+        basedir = qsd_params.get("qsd_basedir")
+        if not basedir:
+            basedir = data_dir.get_data_dir() + "/qsd/%s" % name
```
Comment (YongxueHong) [inline provider/qsd.py:200]:
```suggestion
        basedir = qsd_params.get("qsd_basedir", data_dir.get_data_dir() + "/qsd/%s" % name)
```

**Comment by YongxueHong** [comment]
Hi @qingwangrh 
Do we need the PPC team to help review it?
CC @zhenyzha

**Review by YongxueHong — APPROVED** [review]

File: provider/qsd.py (hunk 13 above)
Comment (qingwangrh) [reply provider/qsd.py]: updated

File: provider/qsd.py (hunk 12 above)
Comment (qingwangrh) [reply provider/qsd.py]: updated

File: provider/qsd.py [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+
+def add_vubp_into_boot(img_name, params, addr=15):
+    """ Add vhost-user-blk-pci device into boot command line """
+    devs = create_vubp_devices(None, img_name, params)
+    cmd = ""
+    for dev in devs:
+        cmd += " %s" % dev.cmdline()
+    if cmd:
+        machine_type = params.get("machine_type", "")
+        if machine_type.startswith("q35"):
```
Comment (zhenyzha) [inline provider/qsd.py:75]: `if machine_type.startswith("q35") or machine_type.startswith("arm64"):`

File: provider/qsd.py [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+    if vm and not vm.devices.get_by_qid(qid):
+        # Declare virtual QSD daemon
+        qsd = qdevices.QDaemonDev("qsd", aobject=qsd_name,
+                                  child_bus=qdevices.QUnixSocketBus(img_sock,
+                                                                    qsd_name))
+        vm.devices.insert(qsd)
+
+    machine_type = params.get("machine_type", "")
+    qbus_type = "PCI"
+    if machine_type.startswith("q35"):
```
Comment (zhenyzha) [inline provider/qsd.py:121]: if machine_type.startswith("q35") or machine_type.startswith("arm64"):

**Comment by zhenyzha** [comment]
> 

Just confirmed that PPC doesn't need to review this patch

File: provider/qsd.py (hunk 15 above)
Comment (qingwangrh) [reply provider/qsd.py]: updated

File: provider/qsd.py (hunk 14 above)
Comment (qingwangrh) [reply provider/qsd.py]: updated

**Pull request merged**